File: common/src/main/java/com/ssblur/scriptor/word/action/ColorAction.java
package com.ssblur.scriptor.word.action;

import com.ssblur.scriptor.block.ScriptorBlocks;
import com.ssblur.scriptor.blockentity.LightBlockEntity;
import com.ssblur.scriptor.entity.ColorfulSheep;
import com.ssblur.scriptor.entity.ScriptorEntities;
import com.ssblur.scriptor.helpers.CustomColors;
import com.ssblur.scriptor.helpers.targetable.EntityTargetable;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import com.ssblur.scriptor.word.descriptor.Descriptor;
import com.ssblur.scriptor.word.descriptor.duration.DurationDescriptor;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.effect.MobEffects;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.MobSpawnType;
import net.minecraft.world.entity.animal.Sheep;
import net.minecraft.world.item.SpawnEggItem;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;

public class ColorAction extends Action {
  @Override
  public void apply(Targetable caster, Targetable targetable, Descriptor[] descriptors) {
    int color = CustomColors.getColor(descriptors);
    if(targetable instanceof EntityTargetable entityTargetable) {
      if(entityTargetable.getTargetEntity() instanceof Sheep sheep) {
        if(sheep instanceof ColorfulSheep colorfulSheep) {
          colorfulSheep.setColor(color);
        } else {
          CompoundTag tag = new CompoundTag();
          sheep.addAdditionalSaveData(tag);
          tag.putInt("scriptor:color", color);

          var colorfulSheep = ScriptorEntities.COLORFUL_SHEEP_TYPE.get().spawn(
            (ServerLevel) sheep.level(),
            tag,
            null,
            sheep.getOnPos(),
            MobSpawnType.CONVERSION,
            true,
            true
          );
          if(colorfulSheep != null) {
            colorfulSheep.load(tag);
            colorfulSheep.setPos(sheep.position());
            colorfulSheep.setXRot(sheep.getXRot());
            colorfulSheep.setYRot(sheep.getYRot());
            sheep.level().addFreshEntity(colorfulSheep);

            sheep.remove(Entity.RemovalReason.DISCARDED);
          }
        }
      }
    }
  }

  @Override
  public Cost cost() { return new Cost(8, COSTTYPE.ADDITIVE); }

}

File: common/src/main/kotlin/com/ssblur/scriptor/word/action/ColorAction.kt
package com.ssblur.scriptor.word.action

import com.ssblur.scriptor.api.word.Action
import com.ssblur.scriptor.api.word.Descriptor
import com.ssblur.scriptor.color.CustomColors.getColor
import com.ssblur.scriptor.color.interfaces.Colorable
import com.ssblur.scriptor.color.interfaces.ColorableBlock
import com.ssblur.scriptor.color.interfaces.ColorableItem
import com.ssblur.scriptor.helpers.ItemTargetableHelper
import com.ssblur.scriptor.helpers.targetable.EntityTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.registry.colorable.ColorableBlockRegistry.get
import com.ssblur.scriptor.registry.colorable.ColorableBlockRegistry.has
import net.minecraft.world.item.ItemStack

class ColorAction: Action() {
  override fun apply(caster: Targetable, targetable: Targetable, descriptors: Array<Descriptor>) {
    val color = getColor(descriptors)
    if (targetable is EntityTargetable) {
      if (targetable.targetEntity is Colorable) {
        (targetable.targetEntity as Colorable).setColor(color)
      }
    }

    val itemTarget = ItemTargetableHelper.getTargetItemStack(targetable)
    if (!itemTarget.isEmpty) {
      if (itemTarget.item is ColorableItem) {
        val itemOut: ItemStack = (itemTarget.item as ColorableItem).setColor(color, itemTarget)!!
        if (!itemOut.isEmpty) {
          itemTarget.count = 0
          ItemTargetableHelper.depositItemStack(targetable, itemOut)
        }
      }
      return
    }

    val blockEntity = targetable.level.getBlockEntity(
      targetable.offsetBlockPos
    )
    if (blockEntity is Colorable) {
      blockEntity.setColor(color)
      return
    }

    val block = targetable.level.getBlockState(targetable.offsetBlockPos).block
    if (block is ColorableBlock) {
      block.setColor(color, targetable.level, targetable.offsetBlockPos)
    } else if (has(block)) {
      get(block)!!.setColor(color, targetable.level, targetable.offsetBlockPos)
    }
  }

  override fun cost() = Cost(8.0, COSTTYPE.ADDITIVE)
}
